Rework attachment field view so that it triggers events

goldenapples opened the pull request:
See #316.

This is a start at rewriting the "attachment" field view as a Backbone view that extends `editAttributeField`, and therefore has access to the actions defined in there. A big change is that it <del>is no longer browserified and</del> only depends on the public variables exposed in `sui.views`. I think this is a simpler approach and provides a better model to point people to for developing additional field types.

Architecturally - still needs some organizational work. I wanted to put it out for general feedback first, though.

goldenapples committed:
Rework attachment attribute field to extend editAttributeField

Just restructures this attribute field model a bit to better follow the
general pattern of Backbone views.

goldenapples committed:
Allow custom fields to trigger callback.

Abstracts triggerCallbacks to its own method, so it can be called from
attribute fields that extend `editAttributeField` but override its
`updateValue` method.

goldenapples committed:
Remove duplicate toggleClass call.

goldenapples committed:
Un-browserify the attachment field view

I can't see any reason to build in an extra 10K of scripts into this
file which are already being included elsewhere on the page just for the
purpose of encapsulation. I tried to remove all the browserify
dependencies and set this field up as a model for how plugin users could
register their own fields in a more typical Backbone view structure. I
like it a lot better.

I am exposing one additional global, the `iDCache` variable, and that's
somewhat intentional - I'd like for filters on updating attachment to
have access to that data to avoid having to make additional ajax calls
to look it up. The variable name and the way in which it's exposed is
pretty ugly though, and I think it would do better as a public method on
the view...

Also, browserify is still processing this file, I'm just loading the
version from `src` rather than the browserified `build` file. It would
be good to come up with an organizational structure that made more sense
for these view files.

goldenapples committed:
Merge branch 'master' into 316-update-events-attachment-field

goldenapples committed:
Move editAttributeField into main browserified bundle

Following @jitendraharpalani's comments, it makes sense for this to be
bundled into the main plugin js file and avoid the extra unnecessary
HTTP request.

It's still architected basically as a standalone view, which I think is
a good thing. If we keep the pattern used for composing views fairly
standard and standalone, it will be easier to give examples for
extending these views. Everything needed to extend a view should already
be accessible through extending the `sui.views.editAttributeField`
field.

goldenapples committed:
Wait to load attachment before triggering update event

Knowing that a field has updated seems less useful than actually
retrieving the information from that field. This moves the event trigger
into the ajax callback, so that it doesn't trigger until all of the
attachment data has been retrieved.

goldenapples committed:
No longer build attachment field separately

We're no longer loading it separately, so it doesn't need to be
browserified separately.

goldenapples committed:
var self was this, not that

goldenapples committed:
Use a class method for attachment cache object

Setting the cache as a cache method and defining getter/setter helper
functions for it makes it possible to get around using the sui.data
fudge that I was using before.

danielbachhuber commented on js/src/field-attachment.js:
```diff
@@ -1,152 +0,0 @@
-// Cache attachment IDs for quicker loading.
-var iDCache = {};
-
-sui.views.editAttributeFieldAttachment = editAttributeField.extend( {
```
What will the removal of this file break?

goldenapples replied:
Nothing. I'm not sure why Github is showing it as a removal, it's really just being moved to `js/src/views/edit-attribute-field-attachment.js` and built into the `shortcode-ui.js` bundle.

goldenapples committed:
Add some sugar functions for working with callbacks

Added getValue() and setValue() instance methods to editAttributeField,
and a getField() class method for finding a field in a collection.

goldenapples committed:
Setting every value to "value" isn't a good idea

Fix rather silly bug.

goldenapples committed:
getValue wasn't designed to be used with other attributes

I specifically wrote a function called getValue() which takes no
arguments and gets a value. Overloading it by allowing it to get other
attributes seems like it breaks consistency. Also, the way I was doing
it just didn't work.
